File: frontend/src/objects/auth/AuthAccount.ts
// Business note: frontend mirror of a backend service object; keep names and fields aligned unless this file is explicitly frontend-only.
import type {
  AdminId,
  AuthUserId,
  CustomerId,
  IsoDateTime,
  MerchantId,
  PersonName,
  RiderId,
  UserRole,
  Username,
} from '@/objects/core/SharedObjects'
import { ROLE } from '@/objects/core/SharedObjects'

type BaseAuthAccount = {
  id: AuthUserId
  username: Username
  displayName: PersonName
  createdAt: IsoDateTime
}

export type CustomerAuthAccount = BaseAuthAccount & {
  role: typeof ROLE.customer
  linkedProfileId: CustomerId
}

export type MerchantAuthAccount = BaseAuthAccount & {
  role: typeof ROLE.merchant
  linkedProfileId?: MerchantId
}

export type RiderAuthAccount = BaseAuthAccount & {
  role: typeof ROLE.rider
  linkedProfileId: RiderId
}

export type AdminAuthAccount = BaseAuthAccount & {
  role: typeof ROLE.admin
  linkedProfileId?: AdminId
}

export type AuthAccount =
  | CustomerAuthAccount
  | MerchantAuthAccount
  | RiderAuthAccount
  | AdminAuthAccount

export type AuthAccountRole = UserRole

File: backend/src/main/scala/services/auth/objects/AuthAccount.scala
package services.auth.objects

// Business note: service-owned domain object; mirror it in the frontend only when it is part of the protocol surface.
import system.objects.given

import io.circe.{Decoder, Encoder}
import io.circe.generic.semiauto.*
import system.objects.*

final case class AuthAccount(
    id: AuthUserId,
    username: Username,
    role: UserRole,
    displayName: PersonName,
    linkedProfileId: Option[EntityId],
    createdAt: IsoDateTime,
)
object AuthAccount:
  given Encoder[AuthAccount] = deriveEncoder
  given Decoder[AuthAccount] = deriveDecoder
